### PR TITLE
Docs (DialogMd data input): addresses confusion with injecting data

### LINKED ISSUE #1947: Documentation is unclear about how to get data into a Dialog.
#### Bug, feature request, or proposal:
Documentation bug.

#### What is the expected behavior?
When I use the MdDialog service to open a dialog with some component, I'd love to be able to pass some configuration into that component.

#### What is the current behavior?
The `dialogRef` exposes `componentInstance`, which allows me to set properties / call functions on the component after construction, however I'm unsure if this is the recommended approach since it is not mentioned in the docs.

#### What are the steps to reproduce?

Read the docs at https://github.com/angular/material2/blob/master/src/lib/dialog/README.md, and notice the example uses a "static" dialog. 

#### What is the use-case or motivation for changing an existing behavior?
It would improve the documentation and reduce the number of future issues.

#### Which versions of Angular, Material, OS, browsers are affected?
Latest /all

#### Is there anything else we should know?
I can submit a PR for improved documentation of an answer from this threads provides sufficient information!

### PR DESCRIPTION
**Pull request by M-a-c**
* Explained that services are a best practice if you need to pass data back and forth, depending on use case.
* Added an example to show how to inject data into a Dialog.
Fixes #1947

### CODE REVIEW

**Comment by M-a-c** [comment]
Npm failing on tests, not sure whats up.

File: src/lib/dialog/dialog.md [hunk 1]
```diff
@@ -29,6 +29,33 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
...
+import {Component, Inject} from '@angular/core';
+import {MdDialog, MD_DIALOG_DATA} from '@angular/material';
+
+@Component({
+  selector: 'dialog-selector',
+  template: 'passed in {{ data }}',
+})
+
+export class DialogName {
+  constructor@Inject(MD_DIALOG_DATA) public data: any) { }
```
Comment (fxck) [inline src/lib/dialog/dialog.md]: missing `(`

Comment (M-a-c) [inline src/lib/dialog/dialog.md]: LOL how on earth? Must have hit the backspace key switching just before commit, that's super weird..

File: src/lib/dialog/dialog.md [hunk 2]
```diff
@@ -29,6 +29,33 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
 in which they are contained. When closing, an optional result value can be provided. This result
 value is forwarded as the result of the `afterClosed` promise. 
 
+### Sharing Data with the Dialog component.
+Depending on what you are doing you might want to share data with your dialog component. Angular has documentation that explains in general how to share data between any components using [`services`](https://angular.io/docs/ts/latest/cookbook/component-communication.html#!#bidirectional-service).
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: I don't think we should talk about services here, because they're a general Angular pattern. We should mention the `data` option instead.

File: src/lib/dialog/dialog.md [hunk 3]
```diff
@@ -29,6 +29,33 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
...
+import {Component, Inject} from '@angular/core';
+import {MdDialog, MD_DIALOG_DATA} from '@angular/material';
+
+@Component({
+  selector: 'dialog-selector',
+  template: 'passed in {{ data }}',
+})
+
+export class DialogName {
+  constructor( @Inject(MD_DIALOG_DATA) public data: any ) { }
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: We don't do spaces between the braces. It should be `constructor(@Inject(MD_DIALOG_DATA) public data: any) { }`.

File: src/lib/dialog/dialog.md (hunk 2 above)
Comment (M-a-c) [inline src/lib/dialog/dialog.md]: Yeah I agree. Should that ever be addressed in Angular Material documentation? Since a lot of the components are interactive, services seem pretty fundamental to the 'passing data problem' I started with angular 3 months ago and I was introduced to it by a few other developers asking why I was doing everything so backwards. I'm glad they brought it up, because in the long run it made my code a lot easier to understand. 

* Is there a spot that we could expose users to services? (lots of us are beginners)

**Comment by M-a-c** [comment]
added some fixes.

File: src/lib/dialog/dialog.md [hunk 4]
```diff
@@ -29,6 +29,37 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
 in which they are contained. When closing, an optional result value can be provided. This result
 value is forwarded as the result of the `afterClosed` promise. 
 
+### Sharing Data with the Dialog component.
+Depending on what you are doing you might want to share data with your dialog component. The dialog component has a data option that you can include in order to pass data from the open to the dialog. 
+
+Passing outside data to your component is as simple as.
+```ts
+import {MdDialog} from '@angular/material';
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: In this example it's pretty obvious where `MdDialog` comes from so you can remove this line and the `private dialog: MdDialog`.

File: src/lib/dialog/dialog.md [hunk 5]
```diff
@@ -29,6 +29,37 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
...
+
+private dialog: MdDialog;
+
+let dialogRef = dialog.open(DialogName, {
+  data:'your data',
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: Add a space between the `:` and the string.

File: src/lib/dialog/dialog.md [hunk 6]
```diff
@@ -29,6 +29,37 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
...
+});
+```
+
+Here is an example component you can pass data to.
+```ts
+import {Component, Inject} from '@angular/core';
+import {MD_DIALOG_DATA} from '@angular/material';
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: You can leave the imports here since the injection is slightly more convoluted.

File: src/lib/dialog/dialog.md [hunk 7]
```diff
@@ -29,6 +29,37 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
...
+
+@Component({
+  selector: 'dialog-selector',
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: Try renaming this to something like `your-dialog` and the class to `YourDialog`.

File: src/lib/dialog/dialog.md [hunk 8]
```diff
@@ -29,6 +29,37 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
 in which they are contained. When closing, an optional result value can be provided. This result
 value is forwarded as the result of the `afterClosed` promise. 
 
+### Sharing Data with the Dialog component.
+Depending on what you are doing you might want to share data with your dialog component. The dialog component has a data option that you can include in order to pass data from the open to the dialog.
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: You should add backticks(\`) around "data" in "The dialog component has a data option".

File: src/lib/dialog/dialog.md (hunk 2 above)
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: I don't think it's Material's responsibility to expose users to services.

File: src/lib/dialog/dialog.md [hunk 9]
```diff
@@ -29,6 +29,33 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
 in which they are contained. When closing, an optional result value can be provided. This result
 value is forwarded as the result of the `afterClosed` promise. 
 
+### Sharing Data with the Dialog component.
+Depending on what you are doing you might want to share data with your dialog component. The dialog component has a `data` option allowing you to pass data to the component.
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: Sorry for all the requests, but can you reword this to "If you want to share data with your dialog, you can use the  `data` option to pass information to the dialog component.". Also if you remove the "Passing outside data to your component is as simple as." sentence the PR will be good to go.

File: src/lib/dialog/dialog.md [hunk 10]
```diff
@@ -29,6 +29,33 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
...
+
+Passing outside data to your component is as simple as.
+```ts
+let dialogRef = dialog.open(DialogName, {
+  data: 'your data',
+});
+```
+
+Here is an example component you can pass data to.
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: Can you change this to "To access the `data` in your dialog component, you have to use the `MD_DIALOG_DATA` injection token:"?

File: src/lib/dialog/dialog.md [hunk 11]
```diff
@@ -29,6 +29,33 @@ Components created via `MdDialog` can _inject_ `MdDialogRef` and use it to close
...
+@Component({
+  selector: 'your-dialog',
+  template: 'passed in {{ data }}',
+})
+
+export class YourDialog {
+  constructor(@Inject(MD_DIALOG_DATA) public data: any) { }
+}
+```
+
```
Comment (crisbeto) [inline src/lib/dialog/dialog.md]: Can you reduce the amount of extra blank lines here? This is mainly for consistency.

File: src/lib/dialog/dialog.md (hunk 6 above)
Comment (M-a-c) [inline src/lib/dialog/dialog.md]: ok

**Comment by M-a-c** [comment]
Yeah your wording is a lot more crisp. I need to work on this my wording before I make more contributions.
Short sweet and to the point.

**Comment by henriquefm** [comment]
Just a small detail: the name of the dialog class in  `dialog.open(DialogName, ...` (line 36) is different from the actual class on line 51, `export class YourDialog`.
